### DISCUSSION TITLE
best practice for a multi tenant APP ?

### DISCUSSION THREAD

**Discussion by xu4wang** [directus/directus#10902] [Archive]
Hi All

I am new to directus, I am planning to use directus for a SaaS product, in a multi tenant scenario.

There will be multiple users(tenants), each of them should only create/view/update/delete their own records via the APP or API.
In a traditional database design, there will be a "tenant_id" for each records.  when accessing the database, the system will retrieve the tenant_id for current user, and filter the records he/she can access.

If I customize the existing redirectus app for this scenario, will it be easy or there will be a lot of work on customizing API/APP?

Any comments will be much appreciated.

BR,Austin

**Comment by azrikahar** [comment]
It can be difficult to point out a best approach as it can also depends on your business requirement/logic, but there's a short write up on it on our docs over here: https://docs.directus.io/getting-started/glossary/#multitenancy

As the write up mentioned, you can use roles as a way to scope the tenants. There's also a recent question on this which may provide more idea: https://github.com/directus/directus/discussions/9682

That way you'll be able to leverage [the granular permissions configuration](https://docs.directus.io/configuration/users-roles-permissions/#configuring-permissions) provided in Directus and use [filter rules with `$CURRENT_ROLE`](https://docs.directus.io/configuration/filter-rules/#dynamic-variables) for role/tenant level permissions or even `$CURRENT_USER`/`$CURRENT_USER.something` for user level permissions.

**Reply by xu4wang** [reply]
@azrikahar thanks for the quick summary.  I'm exploring the links in your answer.

**Comment by alexsam986** [comment]
Hey Austin, 

Your idea sounds amazing! 

As far as my experience with Directus, it is a quite a good choice for multi-tenancy. You will not need to heavily modify the core app, just the built-in tools along with RBAC + filters would be suffice. However, if you think you need strict isolation, it is better to go with separate databases per tenant. https://www.mirrorfly.com/blog/single-tenant-vs-multi-tenant/
